### ISSUE TITLE
gdal_merge produces strange pixels merging tiff images

### ISSUE THREAD

**Issue by Bremen81** [OSGeo/gdal#1827]
[test images.zip](https://github.com/OSGeo/gdal/files/3594694/test.images.zip)
### Expected behavior and actual behavior.
Hi everyone,
I'm using gdal_merge to merge a number of different tiff images in a mosaic. It always go smooth but with the 3 attached images I got strange artifacts.

### Steps to reproduce the problem.
To reproduce the problem use the following command:
gdal_merge.py -of GTiff -o output.tif 1.tif 2.tif 3.tif 

### Operating system
Windows 8 64 bit

### GDAL version and provenance
I'm using gdal 2.2.2 from gdal.org download page. I've tried the same operation with QGIS 3.8.2 which uses gdal 2.4.1 and I get the same artifacts.

May I ask you to limit the use of these images only for the investigation?

Many thanks,
Alessandro Papaleo

**Comment by kannes** [comment]
Hey Alessandro, 

`gdal_merge.py` is a pretty "dumb" tool, that simply stacks the data of the files you pass to it on top of each other. It does not average things or know about human vision color models.

Each of your files has 4 bands, R, G, B and A.

Here is what happens for one of the "strange" pixels, I picked a pink one at about X=0.90856 Y=40.85477.

**`1.tif`:**
- Not covering that area so it is irrelevant.

**`2.tif`:**
- Band 1: 113
- Band 2: 128
- Band 3: 151
- Band 4: 255

**`3.tif`:**
- Band 1: no data
- Band 2: 10
- Band 3: no data
- Band 4: 255

So `gdal_merge.py` now checks which value the pixel should get. https://gdal.org/programs/gdal_merge.html says "In areas of overlap, the last image will be copied over earlier ones." It should say that this is done band-wise, I will add that to the documentation.

In the command line the order was `1.tif` first, `2.tif` second, `3.tif` last.

So for the first band, it checks `3.tif`, finds no value, checks `2.tif` and finds `113`. So the output image's band 1 (R if you interpret it like that) gets the value `113`.

For the second band `3.tif` has the value `10`. The output image's band 2 (G if you interpret it like that) gets the value `10`.

Band 3 again has no value in `3.tif` so the value `151` from `2.tif` is taken. The output image's band 3 (B if you interpret it like that) gets the value `151`.

Band 4 has the value `255` in `3.tif`. The output image's band 4 (A if you interpret it like that) gets the value `255`.

This leads to the following result:

**`output.tif`:**
Band 1: 113
Band 2: 10
Band 3: 151
Band 4: 255

Lots of color in the R and B bands, little in G, that's pink: #710A97

So in conclusion, because `gdal_merge.py` blindly and stupidly tries to fill all bands with values but then we humans look at them as 3-4 part colors (in this case), the result *looks* weird. If those bands were e.g. temperature, wind, elevation, it would be easier to understand.

Hope this helps to understand even if I cannot offer a solution for what you want to do. Please try https://gis.stackexchange.com for that.

**Comment by rouault** [comment]
> "In areas of overlap, the last image will be copied over earlier ones." It should say that this is done band-wise, I will add that to the documentation.

This is slightly more complicated. It does take into account nodata or alpha of the source image, to avoid replacing target pixels by nodata/transparent source pixels, but it doesn't do any alpha blending, and it indeed only considers nodata band per band. Which is the behaviour that would correspond to the UNIFIED_SRC_NODATA=NO mode of gdalwarp (https://gdal.org/api/gdalwarp_cpp.html#_CPPv4N15GDALWarpOptions16papszWarpOptionsE), whereas gdalwarp default is UNIFIED_SRC_NODATA=YES

**Comment by kannes** [comment]
How about this:
"This utility will automatically mosaic a set of images. All the images must be in the same coordinate system and have a matching number of bands, but they may be overlapping, and at different resolutions. In areas of overlap, the last image will be copied over earlier ones. **Nodata/transparency values are considered on a band by band level, i.e. a nodata/transparent pixel on one source band will not set a nodata/transparent value on all bands for the target pixel in the resulting raster nor will it overwrite a valid pixel value.**

**Comment by rouault** [comment]
> How about this:

sounds good
